### PR TITLE
added exception-response to report exceptions as to HTTP response & l…

### PR DESCRIPTION
**Pull request by erez-rabih**
…atency counters

### CODE REVIEW

**Comment by erez-rabih** [comment]
this PR solves https://github.com/clj-commons/iapetos/issues/47

still missing:
1. tests
2. a more elegant way to organize the code (especially the `safe`) function

File: project.clj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-(defproject clj-commons/iapetos "0.1.10"
+(defproject org.clojars.nanit/iapetos "0.1.11"
```
Comment (slipset) [inline project.clj]: This doesn't seem quite right :)

**Review by slipset — COMMENTED** [review]
Looks ok to me, but I'd really appreciate you updating the docs as well to describe the usage of the new option

File: project.clj (hunk 1 above)
Comment (erez-rabih) [reply project.clj]: sorry I wasn't sure if this is going to get any attention so I released it to our private clojars repo
fixed by https://github.com/clj-commons/iapetos/pull/48/commits/9b1b0655af4dc493bdf4e29dd68a61cfc77bfd22

**Comment by erez-rabih** [comment]
@slipset thanks for your review
1. I fixed the project name
2. I added some docs (please review they are enough)

I would be glad to get some guidance on:
1. a more elegant solution to the [with-custom-exception-response function ](https://github.com/clj-commons/iapetos/pull/48/files#diff-0691a488fd4deee8b5977abc70c08525R130)
2. adding tests to cover the new logic

**Comment by erez-rabih** [comment]
@slipset ping :)

**Comment by slipset** [comment]
Hey, @erez-rabih, been thinking about this for a couple of days, and I've come to the conclusion that your implementation is mixing two concerns, one is counting requests of different types (which is apetos job), and the other is returning an appropriate response (which is the job of the application). Hence I would rather like to see an implementation like:

```Clojure
(defn- run-instrumented
  [{:keys [handler] :as options} request]
  (ex/with-exceptions (exception-counter-for options request)
    (let [start-time (System/nanoTime)
          response   (if (:include-exceptions-in-totals options)
                                (try
                                  (handler request)
                                  (catch Exception e
                                     e))
                                  (handler request))
                            
          delta      (- (System/nanoTime) start-time)]
      (->> (ensure-response-map response)
           (record-metrics! options delta request))
      (if (and (instance? Exception response) (:include-exceptions-in-totals options))
          (throw response)
          response))))
```

The code above is not the most beautiful, but it should give you the idea.

**Comment by erez-rabih** [comment]
@slipset thanks for your answer
I agree that my solution mixes two responsibilities in one place, but I don't think that the proposed solution answers the requirements, let me explain:
Having a look at your proposed solution - you're suggesting adding a boolean `:include-exceptions-in-totals` option to the middleware
Let's assume the option is set to true and `(handler request)` throws an exception. In this case `response` will hold the exception itself which will be converted to a `{:status 200}` map [here](https://github.com/clj-commons/iapetos/blob/master/src/iapetos/collector/ring.clj#L87)
The `record-metrics!` function will record a 200 response code and that's not what I meant to be - I wanted it to be a HTTP error code (5XX)
Eventually the exception would be thrown which is good
So unless I misunderstood something going over the flow, the original need this PR was opened for is not satisfied

Let me offer something else which matches the spirit of your suggestion:
1. add a :exception-response-code option to the middleware options 
2. add another case for `ensure-response-map` [here](https://github.com/clj-commons/iapetos/blob/master/src/iapetos/collector/ring.clj#L87) which returns `{:status (:exception-response-code options)}` if response is an instance of exception
3. this way - the recorded metrics would be of the status requested

what do you think?

**Comment by slipset** [comment]
Then I'd suggest the following:
```Clojure
(defn- run-instrumented
  [{:keys [handler] :as options} request]
    (let [start-time (System/nanoTime)
          response   (if (:include-exceptions-in-totals options)
                                (try
                                  (handler request)
                                  (catch Exception e
                                     e))
                                  (handler request))
                            
          delta      (- (System/nanoTime) start-time)]
     
      (if (and (instance? Exception response) (:include-exceptions-in-totals options))
          (ex/with-exceptions (exception-counter-for options request)
            ;; record whatever metrics you want here
             (record-metrics! options delta request {:status 500} ; the status could as you suggest be another entry in `options` 
             (throw response))
           (->> (ensure-response-map response)
                   (record-metrics! options delta request)))))
```

Again, just sketching here...

**Comment by erez-rabih** [comment]
@slipset I think I didn't get to the bottom of your mind, let me try to ask this question:
the thing that is important for you is that in case an exception is raised by the provided handler, the iapetos handler will throw it and won't respond with an arbitrary provided response map?

**Comment by slipset** [comment]
Yes :) And this because we cannot foresee what our clients wish to do with the exception acted it's been counted.

**Comment by erez-rabih** [comment]
ok I opened a new pr https://github.com/clj-commons/iapetos/pull/48 to replace this one

let's discuss the code change and if we agree I will add documentation

**Pull request closed without merging**